packages/list-store/src/use-searchable-facet.tsx

```diff
@@ -0,0 +1,227 @@
+}
+
+interface FacetState extends FacetProps {
+  setSearchValue: (value: string) => void;
+  toggleLetterCategory: (value: string) => void;
+  setSortBy: (value: FacetSortBy) => void;
+  setFilters: (value: Filter[]) => void;
+}
+
+type FacetStore = ReturnType<typeof createFacetStore>;
+
+export const initialState = {
+  searchValue: '',
+  letterCategory: '',
+  sortBy: 'count' as FacetSortBy,
+  filters: [],
+};
+
+export const createFacetStore = (initProps?: Partial<FacetProps>) => {
+  return createStore<FacetState>()(set => ({
+    ...initialState,
+    ...initProps,
+    setSearchValue: (searchValue: string) =>
+      set(() => ({
+        searchValue,
+      })),
+    toggleLetterCategory: (category: string) =>
+      set(state => ({
+        letterCategory: state.letterCategory === category ? '' : category,
+      })),
+    setSortBy: (sortBy: FacetSortBy) =>
+      set(() => ({
+        sortBy,
+      })),
+    setFilters: (filters: Filter[]) =>
+      set(() => ({
+        filters,
+      })),
+  }));
+};
+
+type FacetProviderProps = React.PropsWithChildren<Partial<FacetProps>>;
+
+export const FacetContext = createContext<FacetStore | null>(null);
+
+export function FacetProvider({children, ...props}: FacetProviderProps) {
+  const storeRef = useRef<FacetStore>();
+  if (!storeRef.current) {
+    storeRef.current = createFacetStore(props);
+  }
+  return (
+    <FacetContext.Provider value={storeRef.current}>
+      {children}
+    </FacetContext.Provider>
+  );
+}
+
+interface GetLetterCategoriesProps {
+  filters: SearchableFilter[];
+  letterCategory?: string;
+  searchValue?: string;
+}
+
+export function getLetterCategories({
+  filters,
+  letterCategory,
+  searchValue,
+}: GetLetterCategoriesProps) {
+  const filtersFilteredBySearchValue = filterBySearchValue(
+    filters,
+    searchValue
+  );
+
+  return [
+    ...new Set([
+      // Always include the letter category of the current filter
+      ...(letterCategory ? [letterCategory] : []),
+      ...filtersFilteredBySearchValue
+        .map(filter => filter.letterCategory)
+        .filter(category => !!category),
+    ]),
+  ].sort();
+}
+
+export function extendFiltersWithLetterCategory(filters?: Filter[]) {
+  if (!filters?.length) {
+    return [];
+  }
+  return filters.map(filter => {
+    let letterCategory = '';
+    const name = filter.name?.toString();
+    if (name) {
+      const firstLetter = name.toUpperCase().replace(/[^0-9A-Z]/g, '')[0];
+      if (/^\d+$/.test(firstLetter)) {
+        letterCategory = '[0-9]';
+      } else {
+        letterCategory = firstLetter;
+      }
+    }
+    return {
+      ...filter,
+      name: name || '',
+      letterCategory,
+    };
+  });
+}
+
+interface GetFilteredFiltersProps {
+  filtersWithLetterCategory: SearchableFilter[];
+  searchValue: string;
+  letterCategory: string;
+  sortBy: FacetSortBy;
+}
+
+function filterBySearchValue(
+  filters: SearchableFilter[],
+  searchValue?: string
+) {
+  if (searchValue) {
```

Comment:
To avoid the `if... else...` you could do:

```
if (!searchValue) {
  return filters;
}

return filters.filter(
  filter => filter.name?.toLowerCase().includes(searchValue.toLowerCase())
);
```